prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1105 @@
+					  (&rx_entry->peer_entry), queue);
+			return FI_SUCCESS;
+		}
+	}
+	util_init_rx_entry(rx_entry, iov, desc, iov_count, addr, context, tag,
+			   flags);
+
+	srx->update_func(srx, rx_entry);
+	return rx_entry->peer_entry.srx->peer_ops->start_tag(
+					&rx_entry->peer_entry);
+}
+
+ssize_t util_srx_generic_recv(struct fid_ep *ep_fid, const struct iovec *iov,
+			      void **desc, size_t iov_count, fi_addr_t addr,
+			      void *context, uint64_t flags)
+{
+	struct util_srx_ctx *srx;
+	struct util_rx_entry *rx_entry;
+	struct slist *queue;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	if (flags & FI_MULTI_RECV)
+		return util_generic_mrecv(srx, iov, desc, iov_count, addr,
+					  context, flags);
+
+	assert(iov_count <= srx->iov_limit);
+	addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
+
+	rx_entry = util_remove_unexp_msg(srx, addr);
+	if (!rx_entry) {
+		queue = addr == FI_ADDR_UNSPEC ? &srx->msg_queue :
+			ofi_array_at(&srx->src_recv_queues, addr);
+		rx_entry = util_get_recv_entry(srx, iov, desc, iov_count, addr,
+					       context, 0, 0, flags);
+		if (!rx_entry)
+			return -FI_ENOMEM;
+		slist_insert_tail((struct slist_entry *)
+				  (&rx_entry->peer_entry), queue);
+		return FI_SUCCESS;
+	}
+
+	util_init_rx_entry(rx_entry, iov, desc, iov_count, addr, context, 0,
+			   flags);
+
+	srx->update_func(srx, rx_entry);
+	return rx_entry->peer_entry.srx->peer_ops->start_msg(
+						&rx_entry->peer_entry);
+}
+
+static ssize_t util_srx_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg,
+				uint64_t flags)
+{
+	struct util_srx_ctx *srx;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	return util_srx_generic_recv(ep_fid, msg->msg_iov, msg->desc,
+				     msg->iov_count, msg->addr, msg->context,
+				     flags | srx->rx_msg_flags);
+}
+
+static ssize_t util_srx_recvv(struct fid_ep *ep_fid, const struct iovec *iov,
+			      void **desc, size_t count, fi_addr_t src_addr,
+			      void *context)
+{
+	struct util_srx_ctx *srx;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	return util_srx_generic_recv(ep_fid, iov, desc, count, src_addr,
+				    context, srx->rx_op_flags);
+}
+
+static ssize_t util_srx_recv(struct fid_ep *ep_fid, void *buf, size_t len,
+			     void *desc, fi_addr_t src_addr, void *context)
+{
+	struct iovec iov;
+	struct util_srx_ctx *srx;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	iov.iov_base = buf;
+	iov.iov_len = len;
+
+	return util_srx_generic_recv(ep_fid, &iov, &desc, 1, src_addr, context,
+				     srx->rx_op_flags);
+}
+
+struct fi_ops_msg util_srx_msg_ops = {
+	.size = sizeof(struct fi_ops_msg),
+	.recv = util_srx_recv,
+	.recvv = util_srx_recvv,
+	.recvmsg = util_srx_recvmsg,
+	.send = fi_no_msg_send,
+	.sendv = fi_no_msg_sendv,
+	.sendmsg = fi_no_msg_sendmsg,
+	.inject = fi_no_msg_inject,
+	.senddata = fi_no_msg_senddata,
+	.injectdata = fi_no_msg_injectdata,
```

Comment:
I don't understand how any of these can be exposed to a user above libfabric.  There's no locking provided anywhere